### PR TITLE
Off-heap statistics and compatibility tests

### PR DESCRIPTION
**Pull request by vjuranek**
https://issues.jboss.org/browse/ISPN-8435

### CODE REVIEW

File: server/integration/testsuite/src/test/java/org/infinispan/server/test/cache/container/OffHeapCompatibilityIT.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+/**
+ * Compatibility mode tests with off-heap data container
+ *
+ * @author vjuranek
+ * @since 9.2
+ */
+@RunWith(Arquillian.class)
+@WithRunningServer({@RunningServer(name = OffHeapCompatibilityIT.CONFIG_NAME)})
+public class OffHeapCompatibilityIT {
```
Comment (gustavonalle) [inline server/integration/testsuite/src/test/java/org/infinispan/server/test/cache/container/OffHeapCompatibilityIT.java]: OFF_HEAP and compatibility mode is (no pun intended) incompatible, IMO it should throw an error if someone configured the cache with both.

Comment (vjuranek) [reply server/integration/testsuite/src/test/java/org/infinispan/server/test/cache/container/OffHeapCompatibilityIT.java]: well, it doesn't, it throws exception later on, see [ISPN-8445](https://issues.jboss.org/browse/ISPN-8445). Also, AFAICT the incompatibility is not documented. 
Btw: why these two are not compatible, both store the data as raw bytes, so I'd expect that they can be used together.

Comment (gustavonalle) [reply server/integration/testsuite/src/test/java/org/infinispan/server/test/cache/container/OffHeapCompatibilityIT.java]: > Btw: why these two are not compatible

In order to interop between remote endpoints and embedded, compat mode internally relies on storing content in "unmarshalled form", so in theory even code running inside the server's JVM would be able to "read" objects without having to deal with marshalling. But OFF_HEAP has a hard requirement on storing byte[ ], so there's a conflict here.

>  both store the data as raw bytes, so I'd expect that they can be used together.

+100, an OFF_HEAP cache in the server should be able to work together with Rest. Have you tried to do so without using compat mode? It should work...

Comment (vjuranek) [reply server/integration/testsuite/src/test/java/org/infinispan/server/test/cache/container/OffHeapCompatibilityIT.java]: > OFF_HEAP cache in the server should be able to work together with Rest. Have you tried to do so without using compat mode? It should work...

yes, it works, test for it is actually [part](https://github.com/infinispan/infinispan/pull/5544/files#diff-60226fbe2e1fa9ef1e73352fa706a3e0R140) of this PR

**Comment by vjuranek** [comment]
well, I removed test for compatibility mode with off-heap

File: core/src/test/java/org/infinispan/stats/SingleStatsTest.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+   protected final int EVICTION_MAX_ENTRIES = 3;
+   protected final int TOTAL_ENTRIES = 5;
+   protected StorageType storageType;
+   protected Cache cache;
+   protected Stats stats;
+
+   @Factory
+   public Object[] factory() {
+      return new Object[]{
+            //new SingleStatsTest().withStorage(StorageType.OBJECT),
```
Comment (wburns) [inline core/src/test/java/org/infinispan/stats/SingleStatsTest.java]: Do we want to reenable this?

Comment (vjuranek) [reply core/src/test/java/org/infinispan/stats/SingleStatsTest.java]: yes, thanks! Added also `BINARY` for sure.

File: server/integration/testsuite/src/test/java/org/infinispan/server/test/cache/container/OffHeapCompatibilityIT.java (hunk 1 above)
Comment (wburns) [reply server/integration/testsuite/src/test/java/org/infinispan/server/test/cache/container/OffHeapCompatibilityIT.java]: I have fixed the issue with OFF-HEAP and compatibility. The issue was that the memory configuration validate method was never invoked :(

https://github.com/infinispan/infinispan/pull/5590

**Comment by wburns** [comment]
I would recommend rebasing to get fixes from master to better run CI.

**Comment by vjuranek** [comment]
rebased

**Comment by tristantarrant** [comment]
Requeued on CI

**Comment by vjuranek** [comment]
failures are not related to this PR

**Comment by wburns** [comment]
Integrated into master, thanks @vjuranek !

**Pull request merged**